### ISSUE TITLE
FR: Flag to override server TLS root certificates

### ISSUE THREAD

**Issue by kev-the-dev** [tailscale/tailscale#8069]
### What are you trying to do?

Use tailscale client with a control server which is hosted with a TLS server certificate signed by a trusted root that does not chain up to the normal system CA. For example, a corporate CA that isn't signed by digikey, etc.

### How should we solve this?

Implement environment variable flags to (optionally) set the root CAs used by tailscale cleint to verify the control server. The default will remain to use the "normal" trust root of the system.

### What is the impact of not solving this?

All users of tailscale must use a control server with a TLS certificate signed by something like digikey / letsencrypt. In some cases, users don't wish to trust these major trust root organizations (which have historically had leaks and vulnerabilities) and would prefer to issue their own CA. 

### Anything else?

Another flag to set the ServerName is also desrireable, so the control server can be verified against a different SNI the the one connecting to.

**Comment by bradfitz** [comment]
I don't think we want to deal with this complexity. It'd need to be in all the clients: iOS (how?), Android, three Mac variants, Windows, etc.

I'd rather we just use the OS's configured trust policies and you can configure your managed corp devices as needed.

**Comment by kev-the-dev** [comment]
I understand your concern for adding complexity. It is possible to manage your OS's root CA to achieve something similar, but there are cases where a user would like to use a different trust system for tailscale then for, say, their browser. 

I'm not so familiar with the mobile app implementation, but my understand is that they share the TLS implementation in this repo. Could you briefly look at the PR submitted to see if that would indeed affect both the desktop and mobile distributions as I understand it too?

**Comment by bradfitz** [comment]
Well, mobile users don't generally go set environment variables before they start their apps :)

What is your goal with all this? You only care about Linux/servers and not desktop/mobile users? What's your control plane? Headscale?

**Comment by kev-the-dev** [comment]
Yes the goal is to use Headscale as the control plane, with headscale using a certificate signed by a custom CA signed for a special name like `my-server-1.company-server`

**Comment by bradfitz** [comment]
So why don't you put your custom CA's root on your Linux clients? Do we not do the right thing if you put your roots in e.g. https://github.com/golang/go/blob/go1.20.3/src/crypto/x509/root_linux.go#L8-L22 ?

**Comment by kev-the-dev** [comment]
As mentioned above, my application requirements different TLS authentication for the control server and for other applications on the system like a browser. Therefore we can't change the global root CA on the system. This could perhaps be done by putting tailscale in a container with a custom root CA, though having flags to set CA would be more convenient and is a common option in applications such as curl

**Comment by kev-the-dev** [comment]
You should be able to add the corporate proxy cert to the system root certs per the maintainers comments above. If that's not feasible, you could run with the patch I created in #8070 what allows you to specify a certificate via an environment variable
